File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    cards = [0] + cards
    visited = [False] * len(cards)
    
    i = 1
    groups = []
    while i < len(cards):
        if not visited[i]:
            visited[i] = True
            group = 1
            j = i
            while not visited[cards[j]]:
                visited[cards[j]] = True
                group += 1
                j = cards[j]
            groups.append(group)
        i += 1
    
    if len(groups) < 2:
        return 0 
    else:
        groups.sort()
        return groups[-1] * groups[-2]

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    public int solution(int[] cards) {
        boolean[] visited = new boolean[cards.length];
        List<Integer> groups = new ArrayList<>();
        
        for (int i = 0; i < visited.length; i++) {
            if (!visited[i]) {
                int group = 0;
                int j = i;
                while(!visited[j]) {
                    group++;
                    visited[j] = true;
                    j = cards[j] - 1;
                }
                groups.add(group);
            }
        }
        
        if (groups.size() == 1) {
            return 0;
        }
        
        groups.sort(Comparator.reverseOrder());
        return groups.get(0) * groups.get(1);
        
    }
}